### PR TITLE
Support `$max` field update operator

### LINKED ISSUE #624: Support `$max` field update operator
https://www.mongodb.com/docs/manual/reference/operator/update/max/

### PR DESCRIPTION
**Pull request by noisersup**
# Description
- [x] functionality
  - [x] handle integers
  - [x] handle all scalar types
  - [x] handle dates
  - [x] handle other types
  - [x] make sure that mixed types are handled correctly
  - [x] implement lexicographic order if string
- [x] integration tests
- [x] compat tests

This PR closes #624 .

## Readiness checklist

* [x] I added tests for new functionality or bugfixes.
* [x] I ran `task all`, and it passed.
* [x] I added/updated comments for both exported and unexported top-level declarations (functions, types, etc).
* [ ] I checked comments rendering with `task godocs`.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

File: integration/update_field_test.go [hunk 1]
```diff
@@ -671,6 +671,178 @@ func TestUpdateFieldInc(t *testing.T) {
 	})
 }
 
+func TestUpdateFieldMax(t *testing.T) {
```
Comment (w84thesun) [inline integration/update_field_test.go:674]: Can we move those cases to compat test and remove this one?

**Review by w84thesun — CHANGES_REQUESTED** [review]
Overall good, please fix linter.

File: integration/update_field_test.go (hunk 1 above)
Comment (noisersup) [reply integration/update_field_test.go:674]: Then what's the point of other Tests in this file (like `TestUpdateFieldInc` for an example?

Comment (noisersup) [reply integration/update_field_test.go:674]: I've updated compat tests, but I think that we still need integration tests because compat tests for both handlers are skipped for now because of document comparison.

File: integration/update_field_compat_test.go [hunk 2]
```diff
@@ -54,6 +58,86 @@ func TestUpdateFieldCompatInc(t *testing.T) {
...
+			update: bson.D{{"$max", bson.D{{"v", -54.32}}}},
+		},
+
+		"MultipleQueries": {
+			update: bson.D{{"$max", bson.D{{"v", int32(39)}, {"a", int32(30)}}}},
+		},
+		"DuplicateQuery": {
+			update: bson.D{{"$max", bson.D{{"v", int32(39)}, {"v", int32(30)}}}},
+			skip:   "Handle duplicates correctly",
+		},
```
Comment (noisersup) [inline integration/update_field_compat_test.go:96]:
```
expected: mongo.WriteException(mongo.WriteException{WriteConcernError:(*mongo.WriteConcernError)(nil), WriteErrors:mongo.WriteErrors{mongo.WriteError{Index:0, Code:40, Message:"Updating the path 'v' would create a conflict at 'v'", Details:bson.Raw(nil), Raw:bson.Raw(nil)}}, Labels:[]string(nil), Raw:bson.Raw(nil)})
actual  : mongo.CommandError(mongo.CommandError{Code:0, Message:"connection(127.0.0.1:37525[-4]) socket was unexpectedly closed: EOF", Labels:[]string{"NetworkError", "RetryableWriteError"}, Name:"", Wrapped:topology.ConnectionError{ConnectionID:"127.0.0.1:37525[-4]", Wrapped:(*errors.errorString)(0xc00007a120), init:false, message:"socket was unexpectedly closed"}, Raw:bson.Raw(nil)})
```

Do we have any issue related to that? The closest one I found is #1101.

File: integration/update_field_test.go (hunk 1 above)
Comment (noisersup) [reply integration/update_field_test.go]: Nevermind, integration tests removed.

File: internal/handlers/common/update.go [hunk 3]
```diff
@@ -328,6 +334,45 @@ func processIncFieldExpression(doc *types.Document, updateV any) (bool, error) {
...
+		// if the key is not found, set it
+		if val != nil {
+			res := types.CompareOrder(val, maxVal, types.Ascending)
+			switch res {
+			case types.Equal:
+				fallthrough
+			case types.Greater:
+				continue
+			case types.Less:
+			case types.Incomparable:
```
Comment (noisersup) [inline internal/handlers/common/update.go:362]: Should I leave an empty `types.Less` case or make a nolint here?

File: internal/handlers/common/update.go [hunk 4]
```diff
@@ -328,6 +334,45 @@ func processIncFieldExpression(doc *types.Document, updateV any) (bool, error) {
...
+	for _, field := range maxExpression.Keys() {
+		val, _ := doc.Get(field)
+
+		maxVal, err := maxExpression.Get(field)
+		if err != nil {
+			// if max field does not exist, don't change anything
+			continue
+		}
+
+		// if the key was not found, set it
```
Comment (w84thesun) [inline internal/handlers/common/update.go:353]: The comment doesn’t describe what’s happening in that if statement. Inside that if we are calling Compare and analyse compare results.

File: internal/handlers/common/update.go (hunk 3 above)
Comment (w84thesun) [reply internal/handlers/common/update.go:362]: You could add a comment like “document value less than update one, set the new value”

**Review by w84thesun — CHANGES_REQUESTED** [review]

File: internal/handlers/common/update.go (hunk 4 above)
Comment (noisersup) [reply internal/handlers/common/update.go:353]: But it describes what's happening when condition is not met and the whole motivation behind this if statement. But I can rephrase it to "if the document value was found, compare it with max value"

**Review by w84thesun — APPROVED** [review]

File: internal/types/compare_order.go [hunk 5]
```diff
@@ -64,6 +64,8 @@ func detectDataType(value any) compareTypeOrderResult {
 		return booleanDataType
 	case time.Time:
 		return dateDataType
+	case *Document:
+		return documentDataType
```
Comment (AlekSi) [inline internal/types/compare_order.go:68]: It should be above array to follow our order

File: integration/update_field_compat_test.go (hunk 2 above)
Comment (AlekSi) [reply integration/update_field_compat_test.go:96]: > socket was unexpectedly closed: EOF

The real error is in the log above

File: integration/update_field_compat_test.go [hunk 6]
```diff
@@ -54,6 +58,86 @@ func TestUpdateFieldCompatInc(t *testing.T) {
 	testUpdateCompat(t, testCases)
 }
 
+func TestUpdateFieldCompatMax(t *testing.T) {
+	t.Parallel()
+	setup.SkipForTigrisWithReason(t, "tigris does not support dynamic types")
```
Comment (AlekSi) [inline integration/update_field_compat_test.go:63]: What does it mean and why we skip all tests for Tigris? We should not do that

**Review by AlekSi — CHANGES_REQUESTED** [review]

File: integration/update_field_compat_test.go (hunk 6 above)
Comment (noisersup) [reply integration/update_field_compat_test.go:63]: When you specify a max value in a type other than existing in database, it will compare them and overwrite based on type comparison order. In tigris, as far as I know we cannot set the value with different type than existing (i.e. types are not dynamic)

Comment (AlekSi) [reply integration/update_field_compat_test.go:63]: So we can't test that with Tigris at all and have to skip all tests?

Comment (noisersup) [reply integration/update_field_compat_test.go:63]: Definitely there should be some basic tigris testing. Although I'm not sure if there's any way to modify compat tests to exclude some collections for specified test

Comment (AlekSi) [reply integration/update_field_compat_test.go:63]: https://github.com/FerretDB/FerretDB/blob/be84709aa9d9363bf566485a25ea34006e9aa223/integration/update_compat_test.go#L128-L133

Those tests should already be skipped by that code

File: integration/update_field_compat_test.go (hunk 2 above)
Comment (noisersup) [reply integration/update_field_compat_test.go:96]: > The real error is in the log above
```
[document.go:242 bson.(*Document).ReadFrom] bson.Document.ReadFrom: types.ConvertDocument: types.Document.validate: keys and values count mismatch: 1 != 2
```
I still believe that key duplicates are handled incorrectly

File: integration/update_field_compat_test.go (hunk 6 above)
Comment (noisersup) [reply integration/update_field_compat_test.go]: Thanks, it works!

File: integration/update_field_compat_test.go [hunk 7]
```diff
@@ -54,6 +56,85 @@ func TestUpdateFieldCompatInc(t *testing.T) {
...
+		},
+		"Double": {
+			update: bson.D{{"$max", bson.D{{"v", 54.32}}}},
+		},
+		"DoubleNegative": {
+			update: bson.D{{"$max", bson.D{{"v", -54.32}}}},
+		},
+
+		"MultipleQueries": {
+			update: bson.D{{"$max", bson.D{{"v", int32(39)}, {"a", int32(30)}}}},
```
Comment (rumyantseva) [inline integration/update_field_compat_test.go:88]: Do I understand correctly that a test case where we set `$max` to a field that doesn't exist is covered by this case too?

I mean something like
```
update: bson.D{{"$max", bson.D{{"a", int32(30)}}}},
```

**Review by rumyantseva — COMMENTED** [review]
Looks good to me overall, asked one question.

File: integration/update_field_compat_test.go (hunk 7 above)
Comment (noisersup) [reply integration/update_field_compat_test.go:88]: It looks like it is (we compare modifiedCount and collection values) but to be more detailed I can add another test case. It will be better for debugging. Thanks!

**Comment by noisersup** [comment]
Tigris converts incoming data to underlaying type of existing value in database, for an example when we run [Int32Lower](https://github.com/FerretDB/FerretDB/pull/1124/files#diff-02beb83969c6feb8b48d15aebcd299647e0aa2e8b1e5c531e74eb2c4f636ad85R63-R65) on double-smallest collection, we send the int32 data to the tigris and tigris on the backend converts the int to the float type, which results in failing test:
```
--- expected
+++ actual
@@ -6,3 +6,5 @@
   "_id": "double-smallest",
-  "v": 30
+  "v": {
+    "$f": 30
+  }
```

@w84thesun said that there were some efforts to check the type before sending it to tigris but it wasn't that easy.

@AlekSi Should we skip tests for tigris here?

**Comment by noisersup** [comment]
Test for tigris are not failing locally, there are issues with tigris build

File: integration/update_field_compat_test.go [hunk 8]
```diff
@@ -54,6 +56,97 @@ func TestUpdateFieldCompatInc(t *testing.T) {
...
+			update:        bson.D{{"$max", bson.D{{"v", false}}}},
+			skipForTigris: "fixing tigris validation needed",
+		},
+		"EmptyOperand": {
+			update:     bson.D{{"$max", bson.D{}}},
+			resultType: emptyResult,
+		},
+		"DateTime": {
+			update:        bson.D{{"$max", bson.D{{"v", primitive.NewDateTimeFromTime(time.Date(2021, 11, 1, 12, 18, 42, 123000000, time.UTC))}}}},
+			skipForTigris: "fixing tigris validation needed",
```
Comment (rumyantseva) [inline integration/update_field_compat_test.go:139]: I don't think it could be fixed at all :) 

So, if you want this test to work for Tigris, you must work with a data set where v is defined as a date. Tigris doesn't allow to store multiple types in the same field. 

A possible solution would be to separate those test cases into a few different functions based on data (e.g. one function for bools, another function for doubles, another function for date etc). Each function will call `setup.Setup` for a particular data provider (separate for bools, separate for ints etc). This way it will work for Tigris.

However, such test cases will limit the number of situations we check for Postgres (e.g. for postgres we might want to set a date to the field that stored bools). I believe this should be a separate test case for postgres only (where we  can use "skip for Tigris" and set a particular reason that Tigris doesn't support such data types). 

Let's discuss if everyone agrees with this approach.

Comment (AlekSi) [reply integration/update_field_compat_test.go:139]: > So, if you want this test to work for Tigris, you must work with a data set where v is defined as a date.

Correct. We have that dataset (`DateTimes`).

> Tigris doesn't allow to store multiple types in the same field.

That's also correct.

> A possible solution

Sorry, but for what problem?

Comment (rumyantseva) [reply integration/update_field_compat_test.go:139]: > Sorry, but for what problem?

For the case when we want to check that storing various data types in the same field works correctly for Postgres.

Comment (rumyantseva) [reply integration/update_field_compat_test.go:139]: @AlekSi or I don't understand your question. This is why sometimes I prefer synchronous calls over such discussions.

Comment (noisersup) [reply integration/update_field_compat_test.go:139]: If I understand correctly you wan to move these test cases to integration tests?

Comment (AlekSi) [reply integration/update_field_compat_test.go:139]: That should be handled by `testUpdateCompat`. In fact, there is already a code for that:

https://github.com/noisersup/FerretDB/blob/35b050fd7ce29614d018b2f5fd091e546eb6db42/integration/update_compat_test.go#L128-L133

It might not work, but then we should just™ fix it

Comment (rumyantseva) [reply integration/update_field_compat_test.go:139]: So, to resume:
- if we want to test it with Tigris, we need to split those test cases into groups based on particular data types
- if after that we want to test that storing various data types still works for postgres, we need to add an additional postgres-only test for it.

Comment (noisersup) [reply integration/update_field_compat_test.go:139]: @AlekSi the point is that tigris doesn't return any error in that case, because if it's possible they parse your input to the current existing type

Comment (rumyantseva) [reply integration/update_field_compat_test.go:139]: @noisersup as agreed at the daily sync, please mention this issue https://github.com/FerretDB/FerretDB/issues/1061 in "skipForTigris", and that will be it for this task.

**Review by rumyantseva — APPROVED** [review]
LGTM

**Review by w84thesun — APPROVED** [review]
LGTM

**Comment by rumyantseva** [comment]
@AlekSi in principle, we would like to merge it, but GitHub requires your approval as you had requested changes in this PR.

**Review by AlekSi — APPROVED** [review]
